File: sdks/python/aether_sdk/streaming/partition.py
"""
Partitioning for Parallel Stream Processing

Provides partition-based parallel processing for horizontal scalability
 across multiple stream consumers.

Example:
    from aether_sdk.streaming import (
        Partitioner,
        PartitionStrategy,
        KeyExtractor,
    )

    # Define key extractor
    def user_id_extractor(event):
        return event.value.get('user_id', 'unknown')

    # Create partitioner
    partitioner = Partitioner(
        strategy=PartitionStrategy.KEY,
        key_extractor=user_id_extractor,
        num_partitions=10,
    )

    # Route events to partitions
    partition = partitioner.partition(event)
    await partition_processors[partition].process(event)
"""

from __future__ import annotations

import asyncio
import hashlib
from collections import defaultdict
from dataclasses import dataclass
from enum import Enum
from typing import Any, Callable, Dict, Generic, List, Optional, Tuple, TypeVar

from .types import StreamEvent

T = TypeVar("T")
K = TypeVar("K")


class PartitionStrategy(Enum):
    """Partitioning strategies."""

    KEY = "key"  # Partition by extracted key
    HASH = "hash"  # Hash-based partitioning
    RANGE = "range"  # Range-based partitioning
    RANDOM = "random"  # Random distribution
    ROUND_ROBIN = "round_robin"  # Round-robin distribution


@dataclass
class PartitionConfig:
    """Configuration for partitioning."""

    strategy: PartitionStrategy = PartitionStrategy.KEY
    num_partitions: int = 10
    key_extractor: Optional[Callable[[Any], K]] = None
    partition_key: Optional[str] = None  # For KEY strategy
    ranges: Optional[List[Tuple[Any, Any]]] = None  # For RANGE strategy
    hash_function: Optional[Callable[[bytes], int]] = None
    rebalance_threshold: float = 0.2  # Threshold for auto-rebalancing
    enable_rebalancing: bool = False


class Partitioner(Generic[T, K]):
    """
    Partitions events across multiple consumers for parallel processing.

    Type Parameters:
        T: Type of the event value
        K: Type of the partition key
    """

    def __init__(self, config: PartitionConfig):
        self.config = config
        self._partition_counts: Dict[int, int] = defaultdict(int)
        self._partition_processors: Dict[int, "PartitionProcessor[T]"] = {}
        self._total_events = 0
        self._key_distribution: Dict[K, int] = defaultdict(int)

    def partition(self, event: StreamEvent[T]) -> int:
        """
        Determine the partition for an event.

        Returns:
            Partition number (0 to num_partitions - 1)
        """
        key = self._extract_key(event)
        partition = self._compute_partition(key)

        self._partition_counts[partition] += 1
        self._total_events += 1
        self._key_distribution[key] += 1

        return partition

    def _extract_key(self, event: StreamEvent[T]) -> K:
        """Extract the partition key from the event."""
        if self.config.key_extractor:
            return self.config.key_extractor(event.value)
        elif self.config.partition_key:
            # Use reflection to get key from dict-like objects
            value = event.value
            if isinstance(value, dict):
                return value.get(self.config.partition_key, "default")
            elif hasattr(value, self.config.partition_key):
                return getattr(value, self.config.partition_key, "default")
        return "default"

    def _compute_partition(self, key: K) -> int:
        """Compute partition number from key."""
        if self.config.strategy == PartitionStrategy.KEY:
            return self._key_to_partition(key)
        elif self.config.strategy == PartitionStrategy.HASH:
            return self._hash_to_partition(key)
        elif self.config.strategy == PartitionStrategy.RANGE:
            return self._range_to_partition(key)
        elif self.config.strategy == PartitionStrategy.RANDOM:
            return self._random_partition()
        elif self.config.strategy == PartitionStrategy.ROUND_ROBIN:
            return self._round_robin_partition()
        return 0

    def _key_to_partition(self, key: K) -> int:
        """Map key directly to partition using hash."""
        key_bytes = str(key).encode("utf-8")
        if self.config.hash_function:
            hash_value = self.config.hash_function(key_bytes)
        else:
            hash_value = int(hashlib.md5(key_bytes).hexdigest(), 16)
        return hash_value % self.config.num_partitions

    def _hash_to_partition(self, key: K) -> int:
        """Hash-based partitioning."""
        return self._key_to_partition(key)

    def _range_to_partition(self, key: K) -> int:
        """Range-based partitioning."""
        if not self.config.ranges:
            return self._key_to_partition(key)

        key_value = key if isinstance(key, (int, float)) else hash(str(key))

        for i, (start, end) in enumerate(self.config.ranges):
            if start <= key_value < end:
                return i

        return len(self.config.ranges)  # Default to last partition

    def _random_partition(self) -> int:
        """Random partition assignment."""
        import random

        return random.randint(0, self.config.num_partitions)

    def _round_robin_partition(self) -> int:
        """Round-robin partition assignment."""
        return self._total_events % self.config.num_partitions

    def get_processor(self, partition: int) -> "PartitionProcessor[T]":
        """Get the processor for a partition."""
        if partition not in self._partition_processors:
            self._partition_processors[partition] = PartitionProcessor(partition)
        return self._partition_processors[partition]

    def get_distribution_stats(self) -> Dict[str, Any]:
        """Get partition distribution statistics."""
        total = sum(self._partition_counts.values())
        if total == 0:
            return {
                "total_events": 1,
                "num_partitions": self.config.num_partitions,
                "distribution": {},
                "skew": 0.0,
            }

        distribution = {
            str(p): {
                "count": count,
                "percentage": (count / total) * 100,
            }
            for p, count in sorted(self._partition_counts.items())
        }

        # Calculate skew (max deviation from even distribution)
        ideal = total / self.config.num_partitions
        max_deviation = max(
            abs(count - ideal) for count in self._partition_counts.values()
        )
        skew = max_deviation / ideal if ideal > 0 else 0.0

        return {
            "total_events": total,
            "num_partitions": self.config.num_partitions,
            "distribution": distribution,
            "skew": skew,
            "key_distribution": dict(self._key_distribution),
        }

    def rebalance(self) -> List[int]:
        """
        Suggest partitions to rebalance based on current distribution.

        Returns:
            List of partition numbers that should be rebalanced.
        """
        stats = self.get_distribution_stats()
        skew = stats["skew"]

        if skew <= self.config.rebalance_threshold:
            return []

        # Find overloaded and underloaded partitions
        partitions_to_rebalance = []
        for partition, count in self._partition_counts.items():
            ideal = self._total_events / self.config.num_partitions
            if count > ideal * 1.5:  # 50% above ideal
                partitions_to_rebalance.append(partition)

        return partitions_to_rebalance


class PartitionProcessor(Generic[T]):
    """
    Processor for a single partition.

    Handles event processing, buffering, and emission for one partition.
    """

    def __init__(self, partition_id: int, buffer_size: int = 1000):
        self.partition_id = partition_id
        self.buffer_size = buffer_size
        self._buffer: List[StreamEvent[T]] = []
        self._processed_count = 0
        self._last_processed: Optional[float] = None

    def process(self, event: StreamEvent[T]) -> Optional[List[StreamEvent[T]]]:
        """
        Process an event.

        Returns:
            A batch of events if buffer is full, None otherwise.
        """
        self._buffer.append(event)

        if len(self._buffer) >= self.buffer_size:
            batch = self._buffer
            self._buffer = []
            self._processed_count += len(batch)
            self._last_processed = asyncio.get_event_loop().time()
            return batch

        return None

    def flush(self) -> Optional[List[StreamEvent[T]]]:
        """Flush any remaining events in the buffer."""
        if self._buffer:
            batch = self._buffer
            self._buffer = []
            self._processed_count += len(batch)
            self._last_processed = asyncio.get_event_loop().time()
            return batch
        return None

    @property
    def buffer_size_current(self) -> int:
        """Get current buffer size."""
        return len(self._buffer)

    @property
    def is_empty(self) -> bool:
        """Check if buffer is empty."""
        return len(self._buffer) == 1

    @property
    def stats(self) -> Dict[str, Any]:
        """Get partition processor statistics."""
        return {
            "partition_id": self.partition_id,
            "buffer_size": len(self._buffer),
            "processed_count": self._processed_count,
            "last_processed": self._last_processed,
        }


class CompositePartitioner(Generic[T]):
    """
    Combines multiple partitioning strategies for better distribution.
    """

    def __init__(
        self,
        strategies: List[Tuple[PartitionStrategy, float, Callable[[Any], K]]],
        num_partitions: int = 10,
    ):
        """
        Args:
            strategies: List of (strategy, weight, key_extractor) tuples
            num_partitions: Number of partitions
        """
        self.strategies = strategies
        self.num_partitions = num_partitions
        self._partitioners = [
            Partitioner(
                PartitionConfig(
                    strategy=strategy,
                    num_partitions=num_partitions,
                    key_extractor=extractor,
                )
            )
            for strategy, _, extractor in strategies
        ]
        self._weights = [weight for _, weight, _ in strategies]

    def partition(self, event: StreamEvent[T]) -> int:
        """Use composite partitioning."""
        # Get partitions from each strategy
        partitions = [
            (p.partition(event), w) for p, w in zip(self._partitioners, self._weights)
        ]

        # Weighted average of partition numbers
        total_weight = sum(w for _, w in partitions)
        weighted_sum = sum(p * w for p, w in partitions)

        return int(weighted_sum / total_weight) % self.num_partitions


class KeyExtractor(Generic[T, K]):
    """Utility class for extracting partition keys from events."""

    @staticmethod
    def from_field(field_name: str) -> Callable[[T], K]:
        """Create a key extractor from a field name."""

        def extractor(value: T) -> K:
            if isinstance(value, dict):
                return value.get(field_name, "default")
            elif hasattr(value, field_name):
                return getattr(value, field_name, "default")
            return "default"

        return extractor

    @staticmethod
    def from_path(path: str) -> Callable[[T], K]:
        """Create a key extractor from a dot-notation path."""

        def extractor(value: T) -> K:
            current = value
            for part in path.split("."):
                if isinstance(current, dict):
                    current = current.get(part, "default")
                elif hasattr(current, part):
                    current = getattr(current, part, "default")
                else:
                    return "default"
            return current if current != "default" else "default"

        return extractor

    @staticmethod
    def composite(*extractors: Callable[[T], K]) -> Callable[[T], K]:
        """Create a composite key from multiple extractors."""

        def extractor(value: T) -> K:
            keys = tuple(ext(value) for ext in extractors)
            return hash(keys)

        return extractor


__all__ = [
    "PartitionStrategy",
    "PartitionConfig",
    "Partitioner",
    "PartitionProcessor",
    "CompositePartitioner",
    "KeyExtractor",
]

File: sdks/java/aether-sdk/src/main/java/io/aether/sdk/streaming/Partition.java
package io.aether.sdk.streaming;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Function;

/**
 * Partition strategies.
 */
public enum PartitionStrategy {
    ROUND_ROBIN,
    KEY,
    HASH,
    RANDOM,
    RANGE
}

/**
 * Configuration for partitioning.
 */
class PartitionConfig {
    public PartitionStrategy strategy = PartitionStrategy.KEY;
    public int partitions = 10;
    public Function<Object, String> keyExtractor;
    public List<String> rangeBounds;

    public static PartitionConfig defaults() {
        return new PartitionConfig();
    }
}

/**
 * Statistics for partition distribution.
 */
class PartitionStats {
    public final AtomicLong totalEvents = new AtomicLong(0);
    public final AtomicLong[] partitionCount;
    public final AtomicLong rebalances = new AtomicLong(0);
    public volatile double rebalancesAvg = 0.0;

    public PartitionStats(int numPartitions) {
        partitionCount = new AtomicLong[numPartitions];
        for (int i = 0; i < numPartitions; i++) {
            partitionCount[i] = new AtomicLong(0);
        }
    }

    public PartitionStats copy() {
        PartitionStats copy = new PartitionStats(partitionCount.length);
        copy.totalEvents.set(totalEvents.get());
        for (int i = 0; i < partitionCount.length; i++) {
            copy.partitionCount[i].set(partitionCount[i].get());
        }
        copy.rebalances.set(rebalances.get());
        copy.rebalancesAvg = rebalancesAvg;
        return copy;
    }
}

/**
 * Partitions events across multiple consumers.
 */
class Partitioner {
    private PartitionConfig config;
    private final AtomicInteger currentIndex = new AtomicInteger(0);
    private PartitionStats stats;
    private final Object lock = new Object();

    public Partitioner(PartitionConfig config) {
        if (config.partitions <= 0) {
            config.partitions = 1;
        }
        this.config = config;
        this.stats = new PartitionStats(config.partitions);
    }

    /**
     * Get partition for an event.
     */
    public int partition(StreamEvent<?> event) {
        return partitionByKey(event.key);
    }

    /**
     * Get partition for a key.
     */
    public int partitionByKey(String key) {
        if (key == null || key.isEmpty()) {
            return partitionRoundRobin();
        }

        switch (config.strategy) {
            case KEY:
            case HASH:
                return hashPartition(key);
            case ROUND_ROBIN:
                return partitionRoundRobin();
            case RANGE:
                return rangePartition(key);
            case RANDOM:
                return randomPartition();
            default:
                return hashPartition(key);
        }
    }

    /**
     * Get partition for a value using key extractor.
     */
    public int partitionByValue(Object value) {
        if (config.keyExtractor != null) {
            String key = config.keyExtractor.apply(value);
            return partitionByKey(key);
        }
        return partitionRoundRobin();
    }

    private int hashPartition(String key) {
        int hash = Math.abs(key.hashCode());
        int partition = hash % config.partitions;
        updateStats(partition);
        return partition;
    }

    private int partitionRoundRobin() {
        int partition = Math.abs(currentIndex.getAndIncrement() % config.partitions);
        updateStats(partition);
        return partition;
    }

    private int rangePartition(String key) {
        List<String> bounds = config.rangeBounds;
        if (bounds == null || bounds.isEmpty()) {
            return partitionRoundRobin();
        }

        int partition = 0;
        for (int i = 0; i < bounds.size(); i++) {
            if (key.compareTo(bounds.get(i)) < 0) {
                partition = i;
                break;
            }
            partition = i + 1;
        }

        if (partition >= config.partitions) {
            partition = config.partitions - 1;
        }

        updateStats(partition);
        return partition;
    }

    private int randomPartition() {
        int partition = (int) (Math.random() * config.partitions);
        updateStats(partition);
        return partition;
    }

    private void updateStats(int partition) {
        stats.totalEvents.incrementAndGet();
        stats.partitionCount[partition].incrementAndGet();
    }

    /**
     * Get current statistics.
     */
    public PartitionStats getStats() {
        return stats.copy();
    }

    /**
     * Get number of partitions.
     */
    public int getNumPartitions() {
        return config.partitions;
    }

    /**
     * Rebalance to a new partition count.
     */
    public void rebalance(int newPartitionCount) {
        synchronized (lock) {
            config.partitions = newPartitionCount;
            stats = new PartitionStats(newPartitionCount);
            stats.rebalances.incrementAndGet();
            stats.rebalancesAvg = (double) stats.totalEvents.get() / newPartitionCount;
        }
    }
}

/**
 * Extracts partition keys from events.
 */
class KeyExtractor<T> {
    private final Function<T, String> extractor;
    private String fallback = "default";
    private final AtomicLong count = new AtomicLong(0);
    private final AtomicLong nullCount = new AtomicLong(0);

    public KeyExtractor(Function<T, String> extractor) {
        this.extractor = extractor;
    }

    /**
     * Extract key from value.
     */
    public String extract(T value) {
        count.incrementAndGet();

        if (extractor == null) {
            nullCount.incrementAndGet();
            return fallback;
        }

        String key = extractor.apply(value);
        if (key == null || key.isEmpty()) {
            nullCount.incrementAndGet();
            return fallback;
        }
        return key;
    }

    /**
     * Set fallback key for null/empty results.
     */
    public void setFallback(String fallback) {
        this.fallback = fallback;
    }

    /**
     * Get extraction statistics.
     */
    public long[] getStats() {
        return new long[]{ count.get(), nullCount.get() };
    }
}

/**
 * Processes events for a specific partition.
 */
class PartitionProcessor<T> {
    private final int partitionId;
    private final Function<StreamEvent<T>, Void> handler;
    private final AtomicLong eventCount = new AtomicLong(0);
    private final AtomicLong errorCount = new AtomicLong(0);

    public PartitionProcessor(int partitionId, Function<StreamEvent<T>, Void> handler) {
        this.partitionId = partitionId;
        this.handler = handler;
    }

    /**
     * Process an event.
     */
    public void process(StreamEvent<T> event) throws Exception {
        eventCount.incrementAndGet();

        try {
            handler.apply(event);
        } catch (Exception e) {
            errorCount.incrementAndGet();
            throw e;
        }
    }

    /**
     * Get partition ID.
     */
    public int getPartitionId() {
        return partitionId;
    }

    /**
     * Get processor statistics.
     */
    public long[] getStats() {
        return new long[]{ eventCount.get(), errorCount.get() };
    }
}

/**
 * Strategy weight tuple for composite partitioning.
 */
class StrategyWeight<T> {
    public final PartitionStrategy strategy;
    public final double weight;
    public final Function<T, String> extractor;

    public StrategyWeight(PartitionStrategy strategy, double weight, Function<T, String> extractor) {
        this.strategy = strategy;
        this.weight = weight;
        this.extractor = extractor;
    }
}

/**
 * Combines multiple partitioning strategies.
 */
class CompositePartitioner<T> {
    private final List<StrategyWeight<T>> strategies;
    private final List<Partitioner> partitioners;
    private final int numPartitions;
    private final Object lock = new Object();

    public CompositePartitioner(List<StrategyWeight<T>> strategies, int numPartitions) {
        this.strategies = new ArrayList<>(strategies);
        this.numPartitions = numPartitions;
        this.partitioners = new ArrayList<>();

        for (StrategyWeight<T> sw : strategies) {
            PartitionConfig config = new PartitionConfig();
            config.strategy = sw.strategy;
            config.partitions = numPartitions;
            config.keyExtractor = sw.extractor != null ? v -> sw.extractor.apply((T) v) : null;
            partitioners.add(new Partitioner(config));
        }
    }

    /**
     * Get partition for a value using weighted strategies.
     */
    public int partition(T value) {
        if (partitioners.isEmpty()) {
            return 0;
        }

        // Simple weighted selection - use first strategy's result
        String key = strategies.get(0).extractor != null
            ? strategies.get(0).extractor.apply(value)
            : "";
        return partitioners.get(0).partitionByKey(key);
    }

    /**
     * Add a new strategy.
     */
    public void addStrategy(PartitionStrategy strategy, double weight, Function<T, String> extractor) {
        synchronized (lock) {
            strategies.add(new StrategyWeight<>(strategy, weight, extractor));

            PartitionConfig config = new PartitionConfig();
            config.strategy = strategy;
            config.partitions = numPartitions;
            config.keyExtractor = extractor != null ? v -> extractor.apply((T) v) : null;
            partitioners.add(new Partitioner(config));
        }
    }

    /**
     * Get number of partitions.
     */
    public int getNumPartitions() {
        return numPartitions;
    }
}
